atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/partitioning/MathUtils.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.partitioning;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+// todo(snanda): only works for exact partitions  | Also only works for RF = 3
+public final class MathUtils {
+
+    public static List<TokenRingSlice> generateTokenRingSlices(IdealHostPartition idealHostPartition) {
+        List<TokenRingSlice> result = new ArrayList<>();
+
+        for (CassandraHost hostA: idealHostPartition.setA()) {
+            for (CassandraHost hostB: idealHostPartition.setB()) {
+                for (CassandraHost hostC: idealHostPartition.setC()) {
+                    result.add(ImmutableTokenRingSlice.builder().addHosts(hostA, hostB, hostC).build());
+                }
+            }
+        }
+
+        return result;
+    }
+
+    // todo(snanda): only works for exact partitions  | Also only works for RF = 3
```

Comment:
Let's for now assume that number of shards is either a perfect cube, or a product of three numbers that differ by at most one. The number of hosts in each partition can be anything, but we assume it's mostly balanced, and then we can best-effort it

Reply:
We also eventually don't even need to build this. as we can directly determine which shard something should map to by the indices of hosts